src/battery-check/battery-check.c

```diff
@@ -0,0 +1,125 @@
+/* SPDX-License-Identifier: LGPL-2.1-or-later */
+
+#include <errno.h>
+#include <getopt.h>
+#include <sys/socket.h>
+#include <unistd.h>
+
+#include "battery-util.h"
+#include "build.h"
+#include "constants.h"
+#include "errno-util.h"
+#include "glyph-util.h"
+#include "fd-util.h"
+#include "log.h"
+#include "main-func.h"
+#include "socket-util.h"
+#include "terminal-util.h"
+
+static void help(void) {
+        printf("%s\n\n"
+               "Checks battery level to see whether there's enough charge.\n\n"
+               "   -h --help            Show this help\n"
+               "      --version         Show package version\n",
+               program_invocation_short_name);
+}
+
+static void battery_check_send_plymouth_message(const char *message) {
+        assert(message);
+
+        static const union sockaddr_union sa = PLYMOUTH_SOCKET;
+        _cleanup_close_ int fd = -EBADF;
+        size_t n = strlen(message);
+        ssize_t w = n + 1;
+
+        /* We set SOCK_NONBLOCK here so that we rather drop the
+         * message than wait for plymouth */
+        fd = socket(AF_UNIX, SOCK_STREAM|SOCK_CLOEXEC|SOCK_NONBLOCK, 0);
+        if (fd < 0) {
+                log_warning_errno(errno, "socket() failed: %m");
+                return;
+        }
+
+        if (connect(fd, &sa.sa, SOCKADDR_UN_LEN(sa.un)) < 0) {
+                log_full_errno(IN_SET(errno, EAGAIN, ENOENT) || ERRNO_IS_DISCONNECT(errno) ? LOG_DEBUG : LOG_WARNING, errno, "Connection to plymouth failed: %m");
+                return;
+        }
+
+        if (write(fd, message, n + 1) != w)
```

Comment:
so you're just writing a message to plymouth with no protocol in front of it. this won't work. you need the capital letter and the \002

Reply:
I'm not very clear on this sir. Do you mean I should reformat the message before sending it to plymouth? could you show an example sir? thank you.

Reply:
uh, this was in an earlier version of the patch, no? where did this go?

Reply:
So I gave details here:

https://github.com/systemd/systemd/pull/27830#discussion_r1210727962

When plymouth gets sent a request, the first part of the request has to say what kind of request it is. There are many kinds of requests, one of them is "show-message". To do a show-message request, the first byte has to be a capital M.  immediately after that there needs to be an ascii start-of-text control character '\002'.  Immediately after that, there needs to be a byte saying how long the message is. Immediately after that is the message. So instead of doing
``` 
        if (write(fd, message, n + 1) != w)
```
you'd want to do:

```
        _cleanup_free_ char *request = NULL;
        if (asprintf(&request, "M\002%c%s%n", (int) (strlen(message) + 1), message, &n) < 0)
                return (void) log_oom();
        if (write(fd, request, n + 1) != n +1)
```

Reply:
this also neglects to change the mode to shutdown which might be nice.  Maybe it doesn't matter if we're going to add a new mode for battery in the future

Reply:
> uh, this was in an earlier version of the patch, no? where did this go?

@halfline gave guidance, I thought it was complex at the time, and perhaps would not be required. So I didn't add it to the patch. Now It's clearer, and I will implement fully in a few more pushes, with his guidance.

Reply:
> this also neglects to change the mode to shutdown which might be nice. Maybe it doesn't matter if we're going to add a new mode for battery in the future. Also, you should probably test the change with plymouth! if you don't want to overwrite your main install you could set up a VM and change battery-check to think there's low battery in the vm by just hard coding that in the code perhaps

I'd say we add all that we need to make this awesome. 
I want to make a push that formats the message properly. Once that's sorted, I will move to all other requests that remain. The result of doing that is we have a more awesome feature, and I get to learn more. Thank you for taking your time to review sir.